consensus/src/checkpoint.rs

```diff
@@ -112,47 +117,221 @@ mod tests {
-        let mut c3: ProtocolStateChain = ProtocolStateChain(vec![]);
-        let mut b0: ProtocolState = Default::default();
-        b0.body.consensus_state.blockchain_length = Length(0);
-        b0.body.consensus_state.curr_global_slot = GlobalSlot {
-            slot_number: GlobalSlotNumber(0),
-            slots_per_epoch: Length(7140),
-        };
-        let mut b1: ProtocolState = Default::default();
-        b1.body.consensus_state.blockchain_length = Length(1);
-        b1.body.consensus_state.curr_global_slot = GlobalSlot {
-            slot_number: GlobalSlotNumber(1),
-            slots_per_epoch: Length(7140),
-        };
 
-        c0.push(b0).unwrap();
-        c1.push(b1).unwrap();
+        c0.push(a).unwrap();
+        c1.push(b).unwrap();
         assert_eq!(is_short_range(&c0, &c1).unwrap(), true);
         assert_eq!(is_short_range(&c1, &c0).unwrap(), true);
+    }
 
-        init_checkpoints(&mut genesis);
-        let mut b1: ProtocolState = Default::default();
-        b1.body.consensus_state.blockchain_length = Length(2);
-        b1.body.consensus_state.curr_global_slot = GlobalSlot {
-            slot_number: GlobalSlotNumber(2),
-            slots_per_epoch: Length(7140),
-        };
-        let mut b2: ProtocolState = Default::default();
-        b2.body.consensus_state.blockchain_length = Length(667);
-        b2.body.consensus_state.epoch_count = Length(11);
-        b2.body.consensus_state.curr_global_slot = GlobalSlot {
-            slot_number: GlobalSlotNumber(667),
-            slots_per_epoch: Length(7140),
-        };
-        c3.push(b2).unwrap();
-        assert_eq!(is_short_range(&c3, &c0).unwrap(), false);
-        assert_eq!(is_short_range(&c0, &c3).unwrap(), false);
+    #[test]
+    #[wasm_bindgen_test]
+    fn gen_spot_pair_short_aligned_generates_pairs_of_states_in_short_fork_range() {
+        // Both states will share their staking epoch checkpoints.
+        let mut genesis: ProtocolState = Default::default();
+        init_checkpoints(&mut genesis).unwrap();
+
+        let mut a: ProtocolState = Default::default();
+        let mut b: ProtocolState = Default::default();
+
+        gen_spot(&mut a);
+        gen_spot(&mut b);
+        gen_spot_pair_common_checkpoints(&mut a, &mut b, 0);
+
+        let mut c0: ProtocolStateChain = ProtocolStateChain(vec![]);
+        let mut c1: ProtocolStateChain = ProtocolStateChain(vec![]);
+
+        c0.push(a).unwrap();
+        c1.push(b).unwrap();
+        assert_eq!(is_short_range(&c0, &c1).unwrap(), true);
+        assert_eq!(is_short_range(&c1, &c0).unwrap(), true);
+    }
+
+    #[test]
+    #[wasm_bindgen_test]
+    fn gen_spot_pair_short_misaligned_generates_pairs_of_states_in_short_fork_range() {
+        let mut genesis: ProtocolState = Default::default();
+        init_checkpoints(&mut genesis).unwrap();
+
+        let mut a: ProtocolState = Default::default();
+        let mut b: ProtocolState = Default::default();
+
+        gen_spot(&mut a);
```

Comment:
Why not making gen_spot a struct or trait method of ProtocolState

Reply:
Currently, move gen_spot to common.rs such that can reuse gen_spot in any other test cases.

Reply:
Sry my point is not about reusability, but more on OO design

Reply:
Not sure should we put gen_spot as a trait for ProtocolState.

Reply:
`self` is a `&` reference, so the data it refers to cannot be written
![Screenshot (400)](https://user-images.githubusercontent.com/8589224/148724069-7eb303b9-9c64-4ea2-859d-aeb2e9d8a9ce.png)
![Screenshot (399)](https://user-images.githubusercontent.com/8589224/148724080-a073f31c-451b-4498-80c5-51332d53170d.png)

Reply:
Fixed.